server/src/main/java/org/elasticsearch/common/logging/NodeAndClusterIdConverter.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.logging;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.core.LogEvent;
+import org.apache.logging.log4j.core.config.plugins.Plugin;
+import org.apache.logging.log4j.core.pattern.ConverterKeys;
+import org.apache.logging.log4j.core.pattern.LogEventPatternConverter;
+import org.apache.logging.log4j.core.pattern.PatternConverter;
+import org.apache.lucene.util.CloseableThreadLocal;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.common.util.LazyInitializable;
+
+import java.util.Locale;
+import java.util.concurrent.atomic.AtomicReference;
+
+
+@Plugin(category = PatternConverter.CATEGORY, name = "NodeAndClusterIdConverter")
+@ConverterKeys({"node_and_cluster_id"})
+public final class NodeAndClusterIdConverter extends LogEventPatternConverter implements ClusterStateListener {
+
+    private static LazyInitializable<NodeAndClusterIdConverter, Exception> INSTANCE =
+        new LazyInitializable(() -> new NodeAndClusterIdConverter());
+
+    private final Logger logger = LogManager.getLogger(getClass());
+    private final AtomicReference<String> nodeAndClusterIdsReference = new AtomicReference<>();
+    private final CloseableThreadLocal<String> nodeAndClusterIds = new CloseableThreadLocal();
+
+    /**
+     * Called by log4j2 to initialize this converter.
+     */
+    public static NodeAndClusterIdConverter newInstance(final String[] options) {
+        try {
+            return INSTANCE.getOrCompute();
+        } catch (Exception e) {
+            return null;
+        }
+    }
+
+    public NodeAndClusterIdConverter() {
+        super("NodeName", "node_and_cluster_id");
+    }
+
+    @Override
+    public void format(LogEvent event, StringBuilder toAppendTo) {
+        if (nodeAndClusterIds.get() != null) {
```

Comment:
I think it is worth a comment about how this `ThreadLocal` is a cache for the `AtomicReference`. I do wonder, given the benchmarks that you have, if it is worth having the cache at all. It'd be nice to have some perf data for ARM too, just because I suspect it'd be quite different than x86.

Reply:
The comment is definetly worth putting - will do
As I mentioned in our discussion I am wondering if we need that optimization at all. The microbenchmarks I did for that class show that it won't benefit much (it is actually a bit slower)
Mikrobenchmark code&result:
https://github.com/pgomulka/elasticsearch/commit/dc13f47e45ca64910f50eff2e200194de30e4a5b

Reply:
The thread local field has been removed